### PR TITLE
Add time series classification from scratch example

### PR DESCRIPTION
**Pull request by terrytangyuan**

### CODE REVIEW

File: examples/timeseries_classification_from_scratch.qmd [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+## Plot the model's training and validation loss
+
+```{r}
+plot(history)
+```
+
+TODO: Look into why accuracy is different in our implementation.
+
+We can see how the training accuracy reaches almost 0.95 after 100 epochs. However, by observing the validation accuracy we can see how the network still needs training until it reaches almost 0.97 for both the validation and the training accuracy after 200 epochs. Beyond the 200th epoch, if we continue on training, the validation accuracy will start decreasing while the training accuracy will continue on increasing: the model starts overfitting.
```
Comment (terrytangyuan) [inline examples/timeseries_classification_from_scratch.qmd:173]: There's huge difference in loss and accuracy compared to the original Python implementation.

Our implementation:

```
Epoch 49/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6895 - sparse_categorical_accuracy: 0.5306 - val_loss: 0.6950 - val_sparse_categorical_accuracy: 0.5257
Epoch 50/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6893 - sparse_categorical_accuracy: 0.5212 - val_loss: 0.6965 - val_sparse_categorical_accuracy: 0.5160
Epoch 51/500
90/90 [==============================] - 2s 17ms/step - loss: 0.6887 - sparse_categorical_accuracy: 0.5382 - val_loss: 0.6944 - val_sparse_categorical_accuracy: 0.5229
Epoch 00051: early stopping
```
Python implementation:
```
Epoch 380/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0467 - sparse_categorical_accuracy: 0.9872 - val_loss: 0.0947 - val_sparse_categorical_accuracy: 0.9639
Epoch 381/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0464 - sparse_categorical_accuracy: 0.9885 - val_loss: 0.0922 - val_sparse_categorical_accuracy: 0.9653
Epoch 382/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0449 - sparse_categorical_accuracy: 0.9885 - val_loss: 0.0918 - val_sparse_categorical_accuracy: 0.9639
Epoch 383/500
90/90 [==============================] - 1s 6ms/step - loss: 0.0438 - sparse_categorical_accuracy: 0.9889 - val_loss: 0.0905 - val_sparse_categorical_accuracy: 0.9612
Epoch 00383: early stopping

```

File: examples/timeseries_classification_from_scratch.qmd [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+x_test <- test_df[, -1]
+y_test <- test_df[, 1]
+```
+
+## Visualize the data
+
+Here we visualize one timeseries example for each class in the dataset.
+
+```{r}
+library(ggfortify)
```
Comment (terrytangyuan) [inline examples/timeseries_classification_from_scratch.qmd:44]: I hope this is okay. I used this package I maintain: https://github.com/sinhrks/ggfortify

It's easier to read, and readers don't have to learn ggplot2 syntax, which might be out of scope of this example.

File: examples/timeseries_classification_from_scratch.qmd [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+```{r}
+num_classes <- nrow(unique(y_train))
+```
+
+Now we shuffle the training set because we will be using the `validation_split` option
+later when training.
+
+```{r}
+x_train <- x_train[sample(nrow(x_train)),]
```
Comment (t-kalinowski) [inline examples/timeseries_classification_from_scratch.qmd:71]: It looks like you're shuffling the training data here in such a way that the targets `y` not associated with the inputs `x` anymore. You probably want to generate one random permutation of the row indexing vector and use that for both `x_train` and `y_train`. Or keep inputs and targets as part of the same dataframe at this stage and separate them just before the `fit()` call.

This is probably the source of the model accuracy difference you're observing.

File: examples/timeseries_classification_from_scratch.qmd [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+## Setup
+
+```{r}
+library(tensorflow)
+library(keras)
+set.seed(1234)
+```
+
+```{r, message=FALSE}
+train_df <- readr::read_tsv("https://raw.githubusercontent.com/hfawaz/cd-diagram/master/FordA/FordA_TRAIN.tsv", col_names = FALSE)
```
Comment (t-kalinowski) [inline examples/timeseries_classification_from_scratch.qmd:30]: Can you please use `keras::get_file()` here to avoid downloading the dataset on repeated renders?

Something like: 
```
# where the datasets can be accessed
url <- "https://raw.githubusercontent.com/hfawaz/cd-diagram/master/FordA"

train_df <- "FordA_TRAIN.tsv" %>%
  get_file(., file.path(url, .)) %>%
  readr::read_tsv(col_names = FALSE)
```

File: examples/timeseries_classification_from_scratch.qmd (hunk 2 above)
Comment (t-kalinowski) [reply examples/timeseries_classification_from_scratch.qmd:44]: Looks nice! I think it's fine.

**Review by t-kalinowski — COMMENTED** [review]
Looks nice! I think the source of the accuracy difference is due to shuffling the inputs and targets `x` and `y` in training data. See inline comment. 

Thank you!

File: examples/timeseries_classification_from_scratch.qmd [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+For this specific dataset, the data is already z-normalized: each timeseries sample
+has a mean equal to zero and a standard deviation equal to one. This type of
+normalization is very common for timeseries classification problems, see
+[Bagnall et al. (2016)](https://link.springer.com/article/10.1007/s10618-016-0483-9).
+
+In order to use `sparse_categorical_crossentropy`, we will have to count
+the number of classes beforehand.
+
+```{r}
+num_classes <- nrow(unique(y_train))
```
Comment (t-kalinowski) [inline examples/timeseries_classification_from_scratch.qmd:64]: Calling `unique()` on a dataframe is usually bad practice (it's generally slow, since it casts each row as a string by printing it, and then comparing the printed representation). 

`unique` is great for atomic vectors however. I would change this to `length(unique(y_train[[1]]))` or `length(unique(train_df$y))`.

**Comment by terrytangyuan** [comment]
Thank you! Nice catch! I've addressed the comments.

File: examples/timeseries_classification_from_scratch.qmd [hunk 6]
```diff
@@ -49,109 +50,118 @@ Here we visualize one timeseries example for each class in the dataset.
...
 ```{r}
-input_shape <- dim(x_train)[-1]
+input_shape <- dim(x_train)[-1] # drop batch dim
 input_layer <- layer_input(input_shape)
 
 output_layer <- input_layer %>%
+
   # First convolutional layer
-  layer_conv_1d(filters = 64, kernel_size = 3, padding = "same") %>% 
+  layer_conv_1d(64, 3, padding = "same") %>%
```
Comment (terrytangyuan) [inline examples/timeseries_classification_from_scratch.qmd:123]: This was intentional. I remember I had to explicitly pass these names arguments due to some errors in newer TF version. I’d double check though.

Comment (t-kalinowski) [reply examples/timeseries_classification_from_scratch.qmd:123]: the `layer_*` functions do two things:
1. Instantiate a layer instance
2. Compose the layer instance with the first argument passed to `layer_*()`, return the result of the composition. 

"composition" here can mean calling `.add()` on a sequential model and returning the sequential model, or `__call__()`ing the layer instance and returning the resultant tensor.

If you want the `layer_*()` function to skip step 2 and just return the layer instance, the
first argument `object` needs to be missing or NULL. One way to do that is to supply all the other arguments as named arguments, thus avoiding any positional matching and ensuring that the first argument to the function is missing. 

Note that if you do `inputs %>% layer_conv_1d(64)`, `64` is the 2nd argument to the call, _not_ the first.
